client/extensions/woocommerce/app/products/product-form-variations-row.js

```diff
@@ -1,111 +1,229 @@
+		const { editProductVariation, product, variation } = this.props;
 		editProductVariation( product, variation, { regular_price: e.target.value } );
-	};
+	}
 
-	const setWeight = ( e ) => {
+	setWeight = ( e ) => {
+		const { editProductVariation, product, variation } = this.props;
 		editProductVariation( product, variation, { weight: e.target.value } );
-	};
+	}
 
-	const setDimension = ( e ) => {
+	setDimension = ( e ) => {
+		const { editProductVariation, product, variation } = this.props;
 		const dimensions = { ...variation.dimensions, [ e.target.name ]: e.target.value };
 		editProductVariation( product, variation, { dimensions } );
-	};
+	}
 
-	const setStockQuantity = ( e ) => {
+	setStockQuantity = ( e ) => {
+		const { editProductVariation, product, variation } = this.props;
 		const stock_quantity = Number( e.target.value ) >= 0 ? e.target.value : '';
 		editProductVariation( product, variation, { stock_quantity } );
-	};
+	}
 
-	const showDialog = () => {
+	showDialog = () => {
+		const { variation, onShowDialog } = this.props;
 		onShowDialog( variation.id );
-	};
+	}
+
+	onSelect = ( files ) => {
+		const file = head( files );
+		this.setState( {
+			placeholder: file.preview,
+			transientId: file.ID,
+			isUploading: true,
+		} );
+	}
+
+	onUpload = ( file ) => {
+		const { editProductVariation, product, variation } = this.props;
+		const image = {
+			src: file.URL,
+			id: file.ID,
+		};
+		this.setState( { ...image,
+			transientId: null,
+			isUploading: false,
+		} );
+		editProductVariation( product, variation, { image } );
+	}
+
+	onError = () => {
+		this.setState( {
+			placeholder: null,
+			transientId: null,
+			isUploading: false,
+		} );
+	}
 
-	return (
-		<tr className="products__product-form-variation-row">
-			<td className="products__product-id">
-				<div className="products__product-name-thumb">
-					<div className="products__product-form-variation-image"></div>
-					<span className="products__product-name products__variation-settings-link" onClick={ showDialog }>
-						{ formattedVariationName( variation ) }
-					</span>
-				</div>
-			</td>
-			<td>
-				<FormCurrencyInput noWrap
-					currencySymbolPrefix="$"
-					name="price"
-					value={ variation.regular_price || '' }
-					placeholder="0.00"
-					onChange={ setPrice }
-					size="4"
-				/>
-			</td>
-			<td>
-				<div className="products__product-dimensions-weight">
-					<FormDimensionsInput
-						className="products__product-dimensions-input"
-						unit="in"
-						dimensions={ variation.dimensions }
-						onChange={ setDimension }
-						noWrap
+	removeImage = () => {
+		const { editProductVariation, product, variation } = this.props;
+		this.setState( {
+			placeholder: null,
+			transientId: null,
+			isUploading: false,
+			src: null,
+			id: null,
+		} );
+		editProductVariation( product, variation, { image: {} } );
+	}
+
+	renderImage = () => {
+		const { src, placeholder, isUploading } = this.state;
+
+		let image;
```

Comment:
This is probably my old initialize-all-the-things mindset from C days, but if neither conditional is satisfied below, image will be undefined. Would you consider initializing it to false?